### PR TITLE
chore(ci): extract workflow-debounce logic to a tested script (ADR-006, #2967)

### PR DESCRIPTION
**Pull request by rjmurillo**
## Summary

Extracts the `workflow-debounce` composite action's inline debounce logic out of YAML into a tested Python script, per ADR-006 (no logic in workflow YAML). This is a down-payment on #2967 (ADR-006 workflow burndown).

## What changed

- `debounce.py` (new): the debounce logic. Reads `DELAY_SECONDS`, `WORKFLOW_NAME`, `CONCURRENCY_GROUP` from env, sleeps, computes duration, writes `start`/`end`/`duration` to `GITHUB_OUTPUT` and a `GITHUB_STEP_SUMMARY` block. Pure helpers (`iso_utc`, `build_output_lines`, `build_summary`) are unit-tested; `sleep`/`clock` are injectable.
- `action.yml`: the ~52-line inline `run:` block is now a thin `python3 "$GITHUB_ACTION_PATH/debounce.py"` with env wiring. Step outputs (`start`, `end`, `duration`) are preserved, so the two consuming workflows are unaffected.
- `test_debounce.py` (new): 8 tests covering output lines, summary text, exit codes, and edge cases (positive, negative, edge, CLI).

## Behavior

Unchanged. Same inputs, same outputs, same summary. The extraction is mechanical.

## Validation

- 8 pytest pass locally.
- mypy clean, ruff clean (including the CI E501 ratchet).
- The action YAML parses cleanly.
- No em/en dashes.

## Related Files

The consuming workflows `ai-pr-quality-gate.yml` and `ai-spec-validation.yml` read this action's `start`/`end`/`duration` outputs; both keep working because the output contract is preserved. Neither is modified in this PR.

Refs #2967

### CODE REVIEW

File: .github/actions/workflow-debounce/debounce.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    concurrency_group = env.get("CONCURRENCY_GROUP", "")
+
+    try:
+        delay = int(delay_configured)
+    except ValueError:
+        print(
+            f"error: DELAY_SECONDS must be an integer, got {delay_configured!r}",
+            file=sys.stderr,
+        )
+        return 1
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .github/actions/workflow-debounce/debounce.py:99]: Fixed: negative DELAY_SECONDS is now rejected with a clean exit code before sleeping (validated in a02435712). A limit of 0 is still allowed (valid empty/disabled window).

File: .github/actions/workflow-debounce/debounce.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    start = clock()
+    start_iso = iso_utc(start)
+    print("=== Workflow Debouncing ===")
+    print(f"Workflow: {workflow_name}")
+    print(f"Concurrency Group: {concurrency_group}")
+    print(f"Sleeping for {delay} seconds...")
+
+    sleep(delay)
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .github/actions/workflow-debounce/debounce.py]: Fixed in a02435712: `delay < 0` returns exit 1 with a clear message instead of letting sleep() raise. Smoke-tested: DELAY_SECONDS=-3 -> rc=1.

File: tests/workflows/test_debounce.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    assert slept == [10]
+
+
+# --- run(): negative / edge -----------------------------------------------
+
+
+def test_run_rejects_non_integer_delay(capsys: pytest.CaptureFixture[str]) -> None:
+    rc = debounce.run({"DELAY_SECONDS": "soon"}, sleep=lambda _s: None)
+    assert rc == 1
+    assert "DELAY_SECONDS must be an integer" in capsys.readouterr().err
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/workflows/test_debounce.py]: Added test_run_rejects_negative_delay (asserts rc==1, message, and that sleep is never called). Also added test_run_rejects_non_numeric_delay and test_run_accepts_fractional_delay.

File: .github/actions/workflow-debounce/debounce.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    end = clock()
+    end_iso = iso_utc(end)
+    duration = int(round(end - start))
+
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .github/actions/workflow-debounce/debounce.py:132]: Fixed: duration now floors both endpoints then subtracts (int(end) - int(start)), exactly reproducing the old $((END_EPOCH - START_EPOCH)) from two `date +%s` calls, instead of round().

File: tests/workflows/test_debounce.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+def test_run_rounds_fractional_duration(tmp_path: Path) -> None:
+    out = tmp_path / "output"
+    ticks = iter([0.0, 9.6])
+    debounce.run(
+        {"DELAY_SECONDS": "10", "GITHUB_OUTPUT": str(out)},
+        sleep=lambda _s: None,
+        clock=lambda: next(ticks),
+    )
+    assert "duration=10" in out.read_text(encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/workflows/test_debounce.py]: Updated: the rounding test is now test_run_truncates_fractional_duration (9.6s -> 9) plus test_run_duration_floors_both_endpoints, locking the old integer-seconds contract.

File: .github/actions/workflow-debounce/debounce.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+import time
+from collections.abc import Callable, Mapping
+from datetime import UTC, datetime
+
+
+def iso_utc(epoch_seconds: float) -> str:
+    """Format an epoch timestamp as an ISO 8601 UTC string (second precision)."""
+    return datetime.fromtimestamp(epoch_seconds, tz=UTC).strftime(
+        "%Y-%m-%dT%H:%M:%SZ"
+    )
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .github/actions/workflow-debounce/debounce.py:40]: Fixed in c003bf018: switched to timezone.utc (from datetime.UTC) so the script imports on the runner's system python3, which can predate 3.11. UP017 suppressed with a reason comment.

File: .github/actions/workflow-debounce/debounce.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    concurrency_group = env.get("CONCURRENCY_GROUP", "")
+
+    try:
+        delay = int(delay_configured)
+    except ValueError:
+        print(
+            f"error: DELAY_SECONDS must be an integer, got {delay_configured!r}",
+            file=sys.stderr,
+        )
+        return 1
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .github/actions/workflow-debounce/debounce.py:99]: Fixed in a02435712: parse DELAY_SECONDS as float, so fractional values like 0.5 are accepted (matching the old `sleep "$DELAY_SECONDS"`). Non-numeric and negative still exit 1. Smoke-tested DELAY_SECONDS=0.5 -> sleeps 0.5s.

File: tests/workflows/test_debounce.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    assert slept == [10]
+
+
+# --- run(): negative / edge -----------------------------------------------
+
+
+def test_run_rejects_non_integer_delay(capsys: pytest.CaptureFixture[str]) -> None:
+    rc = debounce.run({"DELAY_SECONDS": "soon"}, sleep=lambda _s: None)
+    assert rc == 1
+    assert "DELAY_SECONDS must be an integer" in capsys.readouterr().err
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/workflows/test_debounce.py]: Fixed in a02435712: the test now asserts a 'must be a number' error for non-numeric input and adds test_run_accepts_fractional_delay for the 0.5 case, matching the float-parse behavior.

File: .github/actions/workflow-debounce/debounce.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+    if delay < 0:
+        print(
+            f"error: DELAY_SECONDS must be non-negative, got {delay}",
+            file=sys.stderr,
+        )
+        return 1
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply .github/actions/workflow-debounce/debounce.py:115]: Fixed in 76fd4625b: added a `math.isfinite(delay)` check before the negative check, so nan/inf/-inf are rejected with exit 1 before sleeping. Smoke-tested: DELAY_SECONDS in {nan, inf, -inf} -> rc=1 'must be finite'. Input validation is now complete (non-numeric, non-finite, negative all rejected; zero and fractional accepted).

File: tests/workflows/test_debounce.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+# --- run(): negative / edge -----------------------------------------------
+
+
+def test_run_rejects_non_numeric_delay(capsys: pytest.CaptureFixture[str]) -> None:
+    rc = debounce.run({"DELAY_SECONDS": "soon"}, sleep=lambda _s: None)
+    assert rc == 1
+    assert "DELAY_SECONDS must be a number" in capsys.readouterr().err
+
+
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply tests/workflows/test_debounce.py:112]: Added test_run_rejects_non_finite_delay, parametrized over nan/inf/-inf/Infinity, asserting rc==1, the 'must be finite' message, and that sleep is never called.
